src/include/storage/write_ahead_log/log_io.h

```diff
@@ -149,11 +162,44 @@ class BufferedLogWriter {
    */
   bool IsBufferFull() { return buffer_size_ == common::Constants::LOG_BUFFER_SIZE; }
 
+  /**
+   * Mark that the BufferedLogWriter is now ready for serialization.
+   *
+   * Currently, BufferedLogWriter is used to serialize both over the disk and over the network.
+   * To avoid copying, we refcount the serialization usages of the BufferedLogWriter.
+   * This function initializes the refcount.
+   *
+   * TODO(WAN): Come up with a better way of "sharing" buffers?
+   */
+  void PrepareForSerialization() {
+    NOISEPAGE_ASSERT(serialize_refcount_.load() == 0, "This buffer is already being serialized.");
+    serialize_refcount_.store(NUM_SERIALIZERS);
```

Comment:
I'm sort of thinking out loud here, but is it always true that a `BufferedLogWriter` will be consumed by all "Serializers"? 
The two specific situations I'm thinking of are the following
- If replication is disable then will this refcount every hit 0?
- If the buffer has read only logs then does the replication manager still consume it? Or does this distinction happen on a more granular level of individual logs?

Reply:
Ok I think I see In `LogSerializerTask::HandFilledBufferToWriter`, if replication is disabled then you add an extra call to to `MarkSerialized` so the count will get down to 0. I think this is a bit not obvious. Is there a way to initialize `NUM_SERIALIZERS` to 1 if replication is disabled? It would prevent it from being a `constexpr` but I think it would be slightly cleaner or more obvious as to what's going on. That way you wouldn't need the extra call to `MarkSerialized` in `LogSerializerTask::HandFilledBufferToWriter`

Reply:
This is a great point. Especially once Tianlei's RetentionPolicy stuff makes it in, the treatment is not quite uniform..

Currently thinking of using `PrepareForSerialization(RetentionPolicy ...)` so that it takes in an enum (disk, disk+network, neither) that sets the refcount accordingly.

Reply:
Changed, let me know what you think?

Reply:
I like it, I think it's a lot cleaner/clearer